File: implementations/fsharp/Agent.fs
namespace AgentPlatform.FSharp

open System
open System.Text.Json
open Azure.AI.OpenAI

type BearerTokenPolicy(token: string) =
    inherit Azure.Core.Pipeline.HttpPipelineSynchronousPolicy()
    override _.OnSendingRequest(message: Azure.Core.HttpMessage) =
        message.Request.Headers.SetValue("Authorization", sprintf "Bearer %s" token)

type Agent(apiKey: string, registry: ToolRegistry, config: AgentConfig, ?endpoint: string, ?jwtToken: string) =
    let options = OpenAIClientOptions()
    do
        match jwtToken with
        | Some t when not (String.IsNullOrEmpty t) ->
            options.AddPolicy(BearerTokenPolicy(t), Azure.Core.HttpPipelinePosition.PerCall)
        | _ -> ()

    let client =
        match endpoint with
        | Some ep when not (String.IsNullOrEmpty ep) ->
            OpenAIClient(Uri(ep), Azure.AzureKeyCredential(apiKey), options)
        | _ ->
            OpenAIClient(apiKey, options)

    let mutable messages : ChatRequestMessage list = []
    let mutable interruptRequested = false

    do
        let systemPrompt =
            "You are a helpful AI assistant. You have access to various tools. " +
            "When asked to perform a task, use the tools to gather information and take actions before answering."
        messages <- [ ChatRequestSystemMessage(systemPrompt) ]

    member _.RequestInterrupt() =
        interruptRequested <- true

    member private _.PrepareApiMessages(msgs: ChatRequestMessage list) : ChatRequestMessage list =
        // Phase 2.2: Shallow copy list for API request payload
        msgs |> List.map id

    member private _.CompressContextIfNeeded(msgs: ChatRequestMessage list) : ChatRequestMessage list =
        // Phase 2.3: Context window protection using F# pipeline
        if msgs.Length <= config.ContextWindowLimit then
            msgs
        else
            printfn "  [Context Window Protection] History size (%d) > limit (%d). Trimming middle history..." msgs.Length config.ContextWindowLimit
            let systemPrompt = msgs.Head
            let recentCount = config.ContextWindowLimit - 3
            let recentMessages = 
                msgs 
                |> List.skip (msgs.Length - recentCount)
                |> List.skipWhile (function :? ChatRequestToolMessage -> true | _ -> false)

            let summaryMsg = ChatRequestSystemMessage(
                sprintf "[System: Previous conversation history was trimmed to fit context window. %d earlier messages summarized.]" (msgs.Length - recentMessages.Length - 1)
            )
            systemPrompt :: summaryMsg :: recentMessages

    member private self.ExecuteApiWithRetry(preparedMessages: ChatRequestMessage list, retryCount: int) : Async<Result<ChatCompletions, string>> =
        async {
            let reqOptions = ChatCompletionsOptions(config.Model, preparedMessages)
            reqOptions.Temperature <- Nullable(0.7f)

            let schemas = registry.GetToolSchemas()
            for schema in schemas do
                reqOptions.Tools.Add(ChatCompletionsFunctionToolDefinition(schema))

            try
                let! resp = client.GetChatCompletionsAsync(reqOptions) |> Async.AwaitTask
                return Ok resp.Value
            with ex ->
                printfn "  [API Error Retry %d/%d] %s" (retryCount + 1) config.MaxRetries ex.Message
                if retryCount >= config.MaxRetries - 1 then
                    return Error ex.Message
                else
                    let delayMs = (int (Math.Pow(2.0, float retryCount))) * 1000
                    do! Async.Sleep delayMs
                    return! self.ExecuteApiWithRetry(preparedMessages, retryCount + 1)
        }

    member self.RunAsync(userInput: string) : Async<TurnResult> =
        async {
            // --- Phase 1: Turn Prologue ---
            printfn "\nUser: %s" userInput
            messages <- messages @ [ ChatRequestUserMessage(userInput) ]

            let mutable apiCalls = 0
            interruptRequested <- false
            let mutable emptyContentRetries = 0
            let mutable turnResult : TurnResult option = None

            // --- Phase 2: Main Conversation Loop ---
            while apiCalls < config.MaxIterations && turnResult.IsNone do
                if interruptRequested then
                    printfn "  [Turn Exit] Turn interrupted by user."
                    turnResult <- Some {
                        FinalResponse = ""
                        Messages = messages
                        ApiCalls = apiCalls
                        Completed = false
                        Failed = false
                        Interrupted = true
                        ExitReason = Interrupted
                        Error = None
                    }
                else
                    apiCalls <- apiCalls + 1

                    // 2.2 & 2.3 Message Preparation and Context Compression via F# Pipeline
                    let preparedMessages = 
                        messages 
                        |> self.PrepareApiMessages 
                        |> self.CompressContextIfNeeded

                    // 2.4 Inner API Retry Loop
                    let! apiResult = self.ExecuteApiWithRetry(preparedMessages, 0)

                    match apiResult with
                    | Error err ->
                        turnResult <- Some {
                            FinalResponse = ""
                            Messages = messages
                            ApiCalls = apiCalls
                            Completed = false
                            Failed = true
                            Interrupted = false
                            ExitReason = ApiError err
                            Error = Some err
                        }
                    | Ok completions when completions.Choices.Count = 0 ->
                        turnResult <- Some {
                            FinalResponse = ""
                            Messages = messages
                            ApiCalls = apiCalls
                            Completed = false
                            Failed = true
                            Interrupted = false
                            ExitReason = NoResponse "No choices returned"
                            Error = Some "No choices returned from LLM"
                        }
                    | Ok completions ->
                        let choice = completions.Choices.[0]
                        let message = choice.Message

                        // 2.6 Tool Call Execution Path
                        if not (isNull message.ToolCalls) && message.ToolCalls.Count > 0 then
                            let assistantMsg = ChatRequestAssistantMessage(message.Content)
                            let registeredNames = registry.GetRegisteredNames() |> Set.ofList

                            let newToolMessages = 
                                message.ToolCalls
                                |> Seq.choose (function
                                    | :? ChatCompletionsFunctionToolCall as fnCall -> Some fnCall
                                    | _ -> None)
                                |> Seq.map (fun fnCall ->
                                    assistantMsg.ToolCalls.Add(fnCall)
                                    let name = fnCall.Name
                                    let callId = fnCall.Id
                                    let argsStr = fnCall.Arguments

                                    if not (registeredNames.Contains(name)) then
                                        let avail = registeredNames |> String.concat ", "
                                        let errStr = sprintf "Error: Tool '%s' is not registered. Available tools: %s" name avail
                                        printfn "  [Tool Validation Error] %s" errStr
                                        ChatRequestToolMessage(errStr, callId)
                                    else
                                        try
                                            use doc = JsonDocument.Parse(if String.IsNullOrEmpty argsStr then "{}" else argsStr)
                                            printfn "  [Tool Execution] %s(%s)" name argsStr
                                            let execResult = registry.ExecuteToolAsync(name, argsStr) |> Async.RunSynchronously
                                            printfn "  [Tool Result] %s" execResult
                                            ChatRequestToolMessage(execResult, callId)
                                        with jsonEx ->
                                            let errStr = sprintf "Error: Invalid JSON arguments for tool '%s': %s" name jsonEx.Message
                                            printfn "  [JSON Parse Error] %s" errStr
                                            ChatRequestToolMessage(errStr, callId)
                                )
                                |> Seq.toList

                            // Append assistant message and tool response messages to canonical history
                            messages <- messages @ (assistantMsg :> ChatRequestMessage :: (newToolMessages |> List.map (fun t -> t :> ChatRequestMessage)))
                        else
                            // 2.7 Final Text Response Path
                            let finalText = if isNull message.Content then "" else message.Content.Trim()

                            if String.IsNullOrEmpty finalText then
                                if emptyContentRetries < 2 then
                                    emptyContentRetries <- emptyContentRetries + 1
                                    printfn "  [Empty Response Recovery] Retrying with prompt nudge..."
                                    messages <- messages @ [ ChatRequestUserMessage("Please provide a complete text response summarizing your answer.") ]
                                else
                                    let text = "(empty response)"
                                    printfn "Assistant: %s" text
                                    messages <- messages @ [ ChatRequestAssistantMessage(text) ]
                                    turnResult <- Some {
                                        FinalResponse = text
                                        Messages = messages
                                        ApiCalls = apiCalls
                                        Completed = true
                                        Failed = false
                                        Interrupted = false
                                        ExitReason = TextResponse text
                                        Error = None
                                    }
                            else
                                printfn "Assistant: %s" finalText
                                messages <- messages @ [ ChatRequestAssistantMessage(finalText) ]

                                // --- Phase 4: Turn Finalization ---
                                turnResult <- Some {
                                    FinalResponse = finalText
                                    Messages = messages
                                    ApiCalls = apiCalls
                                    Completed = true
                                    Failed = false
                                    Interrupted = false
                                    ExitReason = TextResponse finalText
                                    Error = None
                                }

            match turnResult with
            | Some res -> return res
            | None ->
                printfn "  [Turn Exit] Reached max iterations (%d)." config.MaxIterations
                return {
                    FinalResponse = "Reached maximum iteration limit."
                    Messages = messages
                    ApiCalls = apiCalls
                    Completed = false
                    Failed = true
                    Interrupted = false
                    ExitReason = BudgetExhausted
                    Error = Some "Budget exhausted"
                }
        }

File: implementations/csharp/src/Skight.AgentPlatform/Agent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Azure.AI.OpenAI;

namespace Skight.AgentPlatform
{
    public class BearerTokenPolicy : Azure.Core.Pipeline.HttpPipelineSynchronousPolicy
    {
        private readonly string _token;
        public BearerTokenPolicy(string token) => _token = token;
        
        public override void OnSendingRequest(Azure.Core.HttpMessage message)
        {
            message.Request.Headers.SetValue("Authorization", $"Bearer {_token}");
        }
    }

    public class TurnResult
    {
        public string FinalResponse { get; set; } = string.Empty;
        public List<ChatRequestMessage> Messages { get; set; } = new();
        public int ApiCalls { get; set; }
        public bool Completed { get; set; }
        public bool Failed { get; set; }
        public bool Interrupted { get; set; }
        public string ExitReason { get; set; } = "unknown";
        public string? Error { get; set; }
    }

    public class Agent
    {
        private readonly OpenAIClient _client;
        private readonly string _model;
        private readonly List<ChatRequestMessage> _messages = new();
        private readonly ToolRegistry _registry;
        private bool _interruptRequested = false;

        public int MaxIterations { get; set; } = 10;
        public int MaxRetries { get; set; } = 3;
        public int ContextWindowLimit { get; set; } = 30;

        public Agent(string apiKey, ToolRegistry registry, string model = "gpt-4o", string? endpoint = null, string? jwtToken = null)
        {
            var options = new OpenAIClientOptions();
            if (!string.IsNullOrEmpty(jwtToken))
            {
                options.AddPolicy(new BearerTokenPolicy(jwtToken), Azure.Core.HttpPipelinePosition.PerCall);
            }

            if (!string.IsNullOrEmpty(endpoint))
            {
                // Azure OpenAI initialization
                _client = new OpenAIClient(new Uri(endpoint), new Azure.AzureKeyCredential(apiKey), options);
            }
            else
            {
                // Standard OpenAI initialization
                _client = new OpenAIClient(apiKey, options);
            }
            _model = model;
            _registry = registry;
            InitializeSystemPrompt();
        }

        private void InitializeSystemPrompt()
        {
            var systemPrompt = "You are a helpful AI assistant. You have access to various tools. " +
                               "When asked to perform a task, use the tools to gather information and take actions before answering.";
            _messages.Add(new ChatRequestSystemMessage(systemPrompt));
        }

        public void RequestInterrupt()
        {
            _interruptRequested = true;
        }

        private List<ChatRequestMessage> PrepareApiMessages(List<ChatRequestMessage> msgs)
        {
            // Phase 2.2: Shallow copy for API payload
            return new List<ChatRequestMessage>(msgs);
        }

        private List<ChatRequestMessage> CompressContextIfNeeded(List<ChatRequestMessage> msgs)
        {
            // Phase 2.3: Context window protection
            if (msgs.Count <= ContextWindowLimit)
            {
                return msgs;
            }

            Console.WriteLine($"  [Context Window Protection] History size ({msgs.Count}) > limit ({ContextWindowLimit}). Trimming middle history...");

            var systemPrompt = msgs[0];
            int recentCount = ContextWindowLimit - 3;
            var recentMessages = msgs.Skip(msgs.Count - recentCount).ToList();

            while (recentMessages.Count > 0 && recentMessages[0] is ChatRequestToolMessage)
            {
                recentMessages.RemoveAt(0);
            }

            var summaryMsg = new ChatRequestSystemMessage(
                $"[System: Previous conversation history was trimmed to fit context window. {msgs.Count - recentMessages.Count - 1} earlier messages summarized.]"
            );

            var result = new List<ChatRequestMessage> { systemPrompt, summaryMsg };
            result.AddRange(recentMessages);
            return result;
        }

        public async Task<TurnResult> RunAsync(string userInput, Func<List<FunctionDefinition>, List<ChatRequestMessage>, Task<ChatCompletions>>? customLlmCaller = null)
        {
            // --- Phase 1: Turn Prologue ---
            Console.WriteLine($"\nUser: {userInput}");
            _messages.Add(new ChatRequestUserMessage(userInput));

            int apiCalls = 0;
            int emptyContentRetries = 0;

            // --- Phase 2: Main Conversation Loop ---
            while (apiCalls < MaxIterations)
            {
                // 2.1 Pre-API Checks
                if (_interruptRequested)
                {
                    _interruptRequested = false; // Reset after handling
                    Console.WriteLine("  [Turn Exit] Turn interrupted by user.");
                    return new TurnResult
                    {
                        FinalResponse = string.Empty,
                        Messages = _messages,
                        ApiCalls = apiCalls,
                        Interrupted = true,
                        ExitReason = "interrupted"
                    };
                }

                apiCalls++;

                // 2.2 & 2.3 Message Preparation and Context Compression
                var preparedMessages = PrepareApiMessages(_messages);
                preparedMessages = CompressContextIfNeeded(preparedMessages);

                // 2.4 Inner Retry Loop for LLM API Call
                ChatCompletions? completions = null;
                Exception? lastError = null;

                for (int retry = 0; retry < MaxRetries; retry++)
                {
                    try
                    {
                        var toolDefinitions = _registry.GetToolSchemas();
                        if (customLlmCaller != null)
                        {
                            completions = await customLlmCaller(toolDefinitions.ToList(), preparedMessages);
                        }
                        else
                        {
                            var options = new ChatCompletionsOptions(_model, preparedMessages)
                            {
                                Temperature = 0.7f
                            };

                            foreach (var toolDef in toolDefinitions)
                            {
                                options.Tools.Add(new ChatCompletionsFunctionToolDefinition(toolDef));
                            }

                            var response = await _client.GetChatCompletionsAsync(options);
                            completions = response.Value;
                        }
                        break;
                    }
                    catch (Exception ex)
                    {
                        lastError = ex;
                        Console.WriteLine($"  [API Error Retry {retry + 1}/{MaxRetries}] {ex.Message}");
                        if (retry == MaxRetries - 1)
                        {
                            return new TurnResult
                            {
                                FinalResponse = string.Empty,
                                Messages = _messages,
                                ApiCalls = apiCalls,
                                Failed = true,
                                ExitReason = "api_error",
                                Error = ex.Message
                            };
                        }
                        await Task.Delay((int)Math.Pow(2, retry) * 1000);
                    }
                }

                if (completions == null || completions.Choices.Count == 0)
                {
                    return new TurnResult
                    {
                        FinalResponse = string.Empty,
                        Messages = _messages,
                        ApiCalls = apiCalls,
                        Failed = true,
                        ExitReason = "no_response",
                        Error = lastError?.Message ?? "No choices returned."
                    };
                }

                var choice = completions.Choices[0];
                var message = choice.Message;

                // 2.6 Tool Call Execution Path
                if (message.ToolCalls != null && message.ToolCalls.Count > 0)
                {
                    var assistantMessage = new ChatRequestAssistantMessage(message.Content);
                    var registeredSchemas = _registry.GetToolSchemas();
                    var registeredNames = new HashSet<string>(registeredSchemas.Select(s => s.Name));

                    foreach (var toolCall in message.ToolCalls)
                    {
                        if (toolCall is ChatCompletionsFunctionToolCall functionCall)
                        {
                            assistantMessage.ToolCalls.Add(functionCall);
                            var name = functionCall.Name;
                            var callId = functionCall.Id;

                            // Self-correction for unregistered tool
                            if (!registeredNames.Contains(name))
                            {
                                var errorMsg = $"Error: Tool '{name}' is not registered. Available tools: {string.Join(", ", registeredNames)}";
                                Console.WriteLine($"  [Tool Validation Error] {errorMsg}");
                                _messages.Add(new ChatRequestToolMessage(errorMsg, callId));
                                continue;
                            }

                            // Validate JSON arguments
                            try
                            {
                                using var doc = JsonDocument.Parse(string.IsNullOrEmpty(functionCall.Arguments) ? "{}" : functionCall.Arguments);
                            }
                            catch (Exception jsonEx)
                            {
                                var errorMsg = $"Error: Invalid JSON arguments for tool '{name}': {jsonEx.Message}";
                                Console.WriteLine($"  [JSON Parse Error] {errorMsg}");
                                _messages.Add(new ChatRequestToolMessage(errorMsg, callId));
                                continue;
                            }

                            Console.WriteLine($"  [Tool Execution] {name}({functionCall.Arguments})");

                            // Execute tool with runtime exception handling
                            try
                            {
                                var result = await _registry.ExecuteToolAsync(name, functionCall.Arguments);
                                Console.WriteLine($"  [Tool Result] {result}");
                                _messages.Add(new ChatRequestToolMessage(result, callId));
                            }
                            catch (Exception execEx)
                            {
                                var errorResult = $"Error executing tool '{name}': {execEx.Message}";
                                Console.WriteLine($"  [Tool Runtime Error] {errorResult}");
                                _messages.Add(new ChatRequestToolMessage(errorResult, callId));
                            }
                        }
                    }

                    // Insert the assistant message before the tool result messages
                    _messages.Insert(_messages.Count - message.ToolCalls.Count, assistantMessage);
                    continue; // Loop again to send tool results to LLM
                }

                // 2.7 Final Text Response Path
                var finalText = (message.Content ?? string.Empty).Trim();

                // Empty Response Recovery
                if (string.IsNullOrEmpty(finalText))
                {
                    if (emptyContentRetries < 2)
                    {
                        emptyContentRetries++;
                        Console.WriteLine("  [Empty Response Recovery] Retrying with prompt nudge...");
                        _messages.Add(new ChatRequestUserMessage("Please provide a complete text response summarizing your answer."));
                        continue;
                    }
                    else
                    {
                        finalText = "(empty response)";
                    }
                }

                Console.WriteLine($"Assistant: {finalText}");
                _messages.Add(new ChatRequestAssistantMessage(finalText));

                // --- Phase 4: Turn Finalization ---
                return new TurnResult
                {
                    FinalResponse = finalText,
                    Messages = _messages,
                    ApiCalls = apiCalls,
                    Completed = true,
                    ExitReason = "text_response"
                };
            }

            // Exceeded iteration budget
            Console.WriteLine($"  [Turn Exit] Reached max iterations ({MaxIterations}).");
            return new TurnResult
            {
                FinalResponse = "Reached maximum iteration limit.",
                Messages = _messages,
                ApiCalls = apiCalls,
                Failed = true,
                ExitReason = "budget_exhausted"
            };
        }
    }
}
